libyul/optimiser/FullInliner.cpp

```diff
@@ -74,27 +75,82 @@ FullInliner::FullInliner(Block& _ast, NameDispenser& _dispenser, Dialect const&
 
 void FullInliner::run()
 {
-	for (auto& statement: m_ast.statements)
-		if (holds_alternative<Block>(statement))
-			handleBlock({}, std::get<Block>(statement));
-
-	// TODO it might be good to determine a visiting order:
-	// first handle functions that are called from many places.
-
 	// Note that the order of inlining can result in very different code.
 	// Since AST IDs and thus function names depend on whether or not a contract
 	// is compiled together with other source files, a change in AST IDs
 	// should have as little an impact as possible. This is the case
 	// if we handle inlining in source (and thus, for the IR generator,
 	// function name) order.
+	// We use stable_sort below to keep the inlining order of two functions
+	// with the same depth.
+	map<YulString, size_t> depths = callDepths();
+	vector<FunctionDefinition*> functions;
 	for (auto& statement: m_ast.statements)
+		if (holds_alternative<FunctionDefinition>(statement))
+			functions.emplace_back(&std::get<FunctionDefinition>(statement));
+	std::stable_sort(functions.begin(), functions.end(), [depths](
+		FunctionDefinition const* _a,
+		FunctionDefinition const* _b
+	) {
+		return depths.at(_a->name) < depths.at(_b->name);
+	});
+	for (FunctionDefinition* fun: functions)
 	{
-		if (!holds_alternative<FunctionDefinition>(statement))
-			continue;
-		FunctionDefinition& fun = std::get<FunctionDefinition>(statement);
-		handleBlock(fun.name, fun.body);
-		updateCodeSize(fun);
+		handleBlock(fun->name, fun->body);
+		updateCodeSize(*fun);
 	}
+
+	for (auto& statement: m_ast.statements)
+		if (holds_alternative<Block>(statement))
+			handleBlock({}, std::get<Block>(statement));
+}
+
+map<YulString, size_t> FullInliner::callDepths() const
+{
+	CallGraph cg = CallGraphGenerator::callGraph(m_ast);
+
+	// Remove calls to builtin functions.
+	for (auto& call: cg.functionCalls)
+		for (auto it = call.second.begin(); it != call.second.end();)
+			if (m_dialect.builtin(*it))
+				it = call.second.erase(it);
+			else
+				++it;
+	cg.functionCalls.erase(""_yulstring);
+
+	map<YulString, size_t> depths;
+	size_t currentDepth = 0;
+
+	while (true)
+	{
+		vector<YulString> removed;
+		for (auto it = cg.functionCalls.begin(); it != cg.functionCalls.end();)
+		{
+			auto const& [fun, callees] = *it;
+			if (callees.empty())
+			{
+				removed.emplace_back(fun);
+				depths[fun] = currentDepth;
+				it = cg.functionCalls.erase(it);
+			}
+			else
+				++it;
+		}
+
+		for (auto& call: cg.functionCalls)
+			call.second -= removed;
+
+		currentDepth++;
+
+		if (removed.empty())
+			break;
+	}
```

Comment:
Assuming that an empty set (for CallGraph map's `.second`) corresponds to say a starting node `s`. And if we invert the map, then the problem of depth finding gets reduced to finding the longest path to all vertices starting from `s`. I think this may be solved easier. (Abstractly this is just putting all edge weights -1 and finding shortest paths from `s`, but there might be a simple way to do it.)

Reply:
Can you elaborate on this a little? It currently does not sound simpler to me :)

Reply:
What I meant is the following. Construct a directed graph where each vertex is a `FunctionDefinition`. There is an edge from `a -> b` if `b` calls `a` (this is same as the "inverted map" of `CallGraph`) Define an extra vertex `s` and extra edges from `s` to a vertex `v` if `v` does not call any other function (i.e., indegree of `v` is zero)

Let's assign weight `-1` to all the edges. Now, the problem of finding depth is the same as finding the shortest path from `s` to all vertices. Theoretically, Bellman-Ford should solve it. Maybe there is a simpler way since all the weights are the same.

But I guess the current implementation is fine too.